### DISCUSSION TITLE
How to build properly from source with CMake

### DISCUSSION THREAD

**Discussion by andrew11221** [microsoft/onnxruntime#8109] [Other Q&A] [answered]
Due to some constraints, I'm unable to use the build.py file and must use cmake directly to build ORT (cross-compiling for another platform). 

I've tried the following, which seemed successful during compilation: 
```
cd mybuild/
cmake ../cmake/ \
-DCMAKE_TOOLCHAIN_FILE=/home/me/w/my-toolchain.cmake \
-DONNX_CUSTOM_PROTOC_EXECUTABLE=/home/me/w/protoc-3.16.0-linux-x86_64/bin/protoc \
-Donnxruntime_BUILD_UNIT_TESTS=OFF \
-Donnxruntime_BUILD_SHARED_LIB=ON \
-Donnxruntime_MINIMAL_BUILD=ON
```
This generated the libonnxruntime.so library as desired, however I haven't been able to link it to a separate C/C++ test project without getting these errors: 
```
/home/my-stuff/my-ld: /home/me/w/onnxruntime/mybuild/libonnxruntime.so: undefined reference to `MlasGemmFloatKernelAvx512F'
/home/my-stuff/my-ld: /home/me/w/onnxruntime/mybuild/libonnxruntime.so: undefined reference to `MlasGemmU8U8KernelAvx2'
/home/my-stuff/my-ld: /home/me/w/onnxruntime/mybuild/libonnxruntime.so: undefined reference to `void MlasConvDepthwiseKernelAvx2<unsigned char>(unsigned char const* const*, unsigned char, unsigned char const*, unsigned char, int*, unsigned long, unsigned long, unsigned long)'
. . .
```
The list of errors goes on, but they are all undefined references to mlas* symbols. I have managed to run my C++ test project linked with the .so generated by build.py earlier -- it's just not working when I compile with the cmake command above. Does anyone see a flaw with how I'm building ORT?

**Comment by yuslepukhin** [comment, marked as answer]
The script was created to manage the complexity of invoking the CMake. I suggest you simply run the script elsewhere with the appropriate arguments and capture the CMake arguments that are logged on the console. Then use them on your target environment. However, you will still need some python, as it invokes flake8 to check python code etc. You would probably need to explore how to disable those things.

**Comment by tracysh** [comment]
Is the platform that you are cross-compiling to still an x86_64 architecture? I'm guessing that's true because the code is still importing x86_64 functions. cmake\onnxruntime_mlas.cmake selects the x86_64 code based on CMAKE_SYSTEM_PROCESSOR. Is your toolchain file overriding this variable? For the other cross-compiling scenarios (iOS and Android), the processor selection ends up being done through an alternate variable.

**Reply by andrew11221** [reply]
Yes the target platform is also an x86_64, just a different OS.

**Reply by tracysh** [reply]
Is your toolchain file overriding CMAKE_SYSTEM_PROCESSOR? To get things building, you'll need to "set(X86_64, TRUE)" in onnxruntime_mlas.cmake. Or what is CMAKE_SYSTEM_NAME set to?

**Reply by andrew11221** [reply]
No my toolchain file does not override CMAKE_SYSTEM_PROCESSOR. Also my CMAKE_SYSTEM_NAME is one that isn't implemented in the cmake files -- so it falls to all the "else" cases. Theoretically, should it work if I just put `set(x86_64, TRUE)` at the top of the onnxruntime_mlas.cmake file?

**Reply by andrew11221** [reply]
Update: setting CMAKE_SYSTEM_PROCESSOR in my toolchain file did the trick! Thank you
